### PR TITLE
Fix vectorization for generic ops fused with conv2d

### PR DESCRIPTION
**Pull request by pzread**
Due to the issue of transformation mark in codegen strategy [1], split the conv2d decompose and vectorization into two `LinalgSingleTilingExpertPass` stages.

After vectorizing conv2d and generic ops, there are extra transfer ops and temporary buffer (#8569) between due to leading unit dims. Use `OptimizeVectorTransferPass` to remove the unit dims on vector transfer ops to fold them away.  

[1]: The codegen strategy pipeline requires all transformation steps to update the marks on ops; otherwise the following transformations won't apply on the ops with outdated marks. This causes the problem as decompose only updates conv2d ops, so the following vectorization won't apply on the other generic ops.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 1]
```diff
@@ -401,6 +399,26 @@ void addConvTileAndDecomposeExpertPassPipeline(OpPassManager &passManager) {
     nestedModulePM.addNestedPass<func::FuncOp>(createCSEPass());
   }
 
+  // Add the sandbox single tiling expert to vectorize.
+  // We can't do the vectorization in the tiling expert above due to an issue in
+  // codegen strategy pipeline. Since we are moving to the transform dialect, we
+  // choose to have a workaround here by splitting them into two stages.
+  {
+    LinalgSingleTilingExpertPassOptions options;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:407]: How do we know this doesnt tile again with the first level of tiling.....

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Want to dig a bit deeper into the split here.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 1 above)
Comment (pzread) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:407]: By default `tilingLevel = -1` and `tileSizes = []` in `LinalgSingleTilingExpertPassOptions`. Therefore the tile transformation in LinalgSingleTilingExpertPass will be disabled. It seems to be confusing tho...

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:407]: The ops won't get tiled if `tilingLevel` is not specified.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/eliminate_vector_transfer_ops.mlir [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// RUN: iree-opt %s --iree-llvmcpu-eliminate-vector-transfer-pass --cse -canonicalize --split-input-file | FileCheck %s
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/eliminate_vector_transfer_ops.mlir:1]: do we need `cse` and `canonicalize`?

**Review by hanhanW — COMMENTED** [review]
Can we use OptimizeVectorTransferPass? It's a common pass and it includes more optimization.

**Comment by pzread** [comment]
> Can we use OptimizeVectorTransferPass? It's a common pass and it includes more optimization.

Yes, we can. Done.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/eliminate_vector_transfer_ops.mlir (hunk 2 above)
Comment (pzread) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/eliminate_vector_transfer_ops.mlir]: Removed this file.

**Review by hanhanW — APPROVED** [review]
The decomposed stage only applies to conv ops. It advances conv ops marker but not other ops.

Splitting the stages looks fine to me. This aligns with transform dialect, since we're doing something like inserting TileOp, PadOp, VectorizeOp, etc. We'll have to break down these stages when migrating to transform dialect. The PR looks reasonable and addresses the vectorization issue.

**Comment by MaheshRavishankar** [comment]
> The decomposed stage only applies to conv ops. It advances conv ops marker but not other ops.
> 
> Splitting the stages looks fine to me. This aligns with transform dialect, since we're doing something like inserting TileOp, PadOp, VectorizeOp, etc. We'll have to break down these stages when migrating to transform dialect. The PR looks reasonable and addresses the vectorization issue.

Dont we use `optimizeVectorTransfers` after bufferization?

**Review by MaheshRavishankar — APPROVED** [review]
Unblocking since my comment was addressed (still have a question about where the optimize vector transfers is inserted here.

**Comment by hanhanW** [comment]
> Dont we use optimizeVectorTransfers after bufferization?

I think the pass was created when most of vector ops work on memrefs. The patterns handle vector on tensors today, so I think it's fine to add it before bufferization.

**Comment by pzread** [comment]
The conv dispatches use `addConvTileAndDecomposeExpertPassPipeline` pipeline and right now it doesn't have `OptimizeVectorTransferPass`. I tested and that pass worked either before the bufferization or after the bufferization.

To be consistent with other CPU pipelines, I moved it after the bufferization, but doing this optimization before the bufferization makes more sense to me. Filed #9201 to track it.

**Comment by pzread** [comment]
The conv dispatches use `addConvTileAndDecomposeExpertPassPipeline` pipeline and right now it doesn't have `OptimizeVectorTransferPass`. I tested and that pass worked either before the bufferization or after the bufferization.

To be consistent with other CPU pipelines, I moved it after the bufferization, but doing this optimization before the bufferization makes more sense to me. Filed #9201 to track it.

**Pull request merged**